data/src/main/java/com/microsoft/azure/kusto/data/auth/CloudDependentTokenProviderBase.java

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package com.microsoft.azure.kusto.data.auth;
+
+import com.microsoft.azure.kusto.data.exceptions.DataClientException;
+import com.microsoft.azure.kusto.data.exceptions.DataServiceException;
+import java.net.URISyntaxException;
+import java.util.HashSet;
+import java.util.Set;
+import org.jetbrains.annotations.NotNull;
+
+public abstract class CloudDependentTokenProviderBase extends TokenProviderBase {
+    private static final String ERROR_INVALID_SERVICE_RESOURCE_URL =
+            "Error determining scope due to invalid Kusto Service Resource URL";
+    protected final Set<String> scopes = new HashSet<>();
+    private boolean initialized = false;
+
+    CloudDependentTokenProviderBase(@NotNull String clusterUrl) throws URISyntaxException {
+        super(clusterUrl);
+    }
+
+    synchronized void initialize() throws DataClientException, DataServiceException {
+        if (initialized) {
+            return;
+        }
+
+        initializeWithCloudInfo(CloudInfo.retrieveCloudInfoForCluster(clusterUrl));
+        initialized = true;
+    }
+
+    protected void initializeWithCloudInfo(CloudInfo cloudInfo) throws DataClientException, DataServiceException {
+        try {
+            scopes.add(cloudInfo.determineScope());
+        } catch (URISyntaxException e) {
+            throw new DataServiceException(clusterUrl, ERROR_INVALID_SERVICE_RESOURCE_URL, e, true);
+        }
+    }
+
+    @Override
+    public String acquireAccessToken() throws DataServiceException, DataClientException {
+        initialize();
+        return acquireAccessTokenAfterCloudInfo();
+    }
+
+    public abstract String acquireAccessTokenAfterCloudInfo() throws DataServiceException, DataClientException;
```

Comment:
I don't love the naming, `acquireAccessTokenAfterCloudInfo()`. I get the issue - `acquireAccessToken()` is used by non-CloudInfo classes as well, so its name should be that.

I don't have any good suggestions. I'm leaning towards:
1) Rename to `acquireAccessTokenAfterCloudEnriched()`
2) Striking this method altogether, and expect subclasses to implement `acquireAccessToken()` by first calling `super()`, and then acquiring the token.